File: app/src/main/java/com/example/mhdstuff/parsing/types/Diversion.java
package com.example.mhdstuff.parsing.types;

import com.example.mhdstuff.parsing.storage.LineStorage;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.List;

public record Diversion(String title, Location location, DateTime from, DateTime to,
                        String publicText, List<LineAlias> lines) {

    public static List<Diversion> parseDiversions(JsonArray array, LineStorage storage) {
        return TypeHelper.parseList(array, (o) -> parse(o, storage));
    }

    public static Diversion parse(JsonObject obj, LineStorage storage) {
        if (obj == null) return null;

        String title = obj.get("title").getAsString();
        Location location = Location.parse(obj);
        DateTime from = DateTime.parse(obj.get("from").getAsString());
        DateTime to = DateTime.parse(obj.get("to").getAsString());

        String publicText = obj.get("content").getAsString();

        List<LineAlias> lines = new ArrayList<>();

        for (JsonElement element : obj.get("lines").getAsJsonArray()) {
            lines.add(LineAlias.parse(element.getAsString(), storage));
        }

        return new Diversion(title, location, from, to, publicText, lines);
    }
}

File: app/src/main/java/io/github/mirancz/libreinfo/parsing/types/Diversion.kt
package io.github.mirancz.libreinfo.parsing.types

import android.os.Parcelable
import com.google.gson.JsonObject
import io.github.mirancz.libreinfo.parsing.storage.LineStorage
import kotlinx.parcelize.Parcelize

/**
 * This class is not [Serializable], see [io.github.mirancz.libreinfo.parsing.types.dto.DiversionDTO] instead
 */
@Parcelize
data class Diversion(
    val id: Int?,
    val title: String,
    val content: String,
    val from: DateTime,
    val to: DateTime,
    val lines: List<LineAlias>?
) : Parcelable {

}


